### PR TITLE
feat: initial implementation of decoder searches.

### PR DESCRIPTION
**Pull request by drugilsberg**
@niladell @jannisborn 

I created three separate classes to implement decoding given logits: GreedySearch, SamplingSearch, BeamSearch.
I also added relatred unit tests.

The beam search is far from being optimal but given the small vocabulary size and sequence length it might be a decent implementation. Especially if we consider it will be used only in inference.
**Please carefully check the implementation the test is really dumb and I don't trust it.**

The PR is only a draft since we have to integrate the search in the generation from the latent space. Once we are convinced the implementations are fine we can proceed with the integration.

### CODE REVIEW

**Comment by drugilsberg** [comment]
An addendum, you will notice there is no concept of beam size. The idea here is that given the limited size of the vocabulary we can explore the full space of possible sequences.

**Comment by drugilsberg** [comment]
Working on havin a step-wise version of all the searches after the discussion had today with @jannisborn.

**Comment by drugilsberg** [comment]
@niladell @jannisborn
Updated PR with `step` method for all the searches to be used in a token-by-token setup. Only things left are:
- check the tests and the implementation (I did this in a rush and I might have overlooked something)
- integrate them in the decoder for the generation, should be trivial following the tests.

**Comment by jannisborn** [comment]
Great work @drugilsberg, looks good to me. I checked the tests.

A few comments:
- I think we should either set the softmax dimensions in all the searches to -1 instead of 1/2 (depending on the autoregressive/one-shot implementation) or perform checks that the sizes are okay. I went for the second and implemented size and type checks in the base class that are executed at every forward call (pushed the code yesterday).
- implemented a test for temperature in the beam search 

I checked for the speed performance with *conservative* but realstic parameters:
```
Vocab size: 100, BatchSize: 16, SeqLen: 200
Beam: Elapsed time 21.203482151031494
.Beam step: Elapsed time 21.050029039382935
.Greedy: Elapsed time 0.0022139549255371094
.Greedy step: Elapsed time 0.002905130386352539
.Sampling: Elapsed time 0.006699085235595703
.Sampling step: Elapsed time 0.05808281898498535
```
As it scales linear with the batch size this becomes slow very quickly.

I therefore introduced the concept of `top_tokens`. This is another user-controllable hyperparameter that defaults to 5. We sort the logits for each time step and only pick the `top_tokens` dictionary tokens. This accelerates almost by an order of magnitude and will accelerate even more the bigger the dictionary becomes:

```
Vocab size: 100, BatchSize: 16, SeqLen: 200
Beam: Elapsed time 2.8716909885406494
.Beam step: Elapsed time 2.996267080307007
.Greedy: Elapsed time 0.0022428035736083984
.Greedy step: Elapsed time 0.0028438568115234375
.Sampling: Elapsed time 0.0061681270599365234
.Sampling step: Elapsed time 0.058073997497558594
```
But of course, it's a proxy and it may have a similar effect like a lower temperature, i.e. less creative generations.
If we would want to keep the option to expand all sequences, we could rename `top_tokens` to `top_token_fraction`, set the default to 1., i.e. keeping all tokens.

**Comment by drugilsberg** [comment]
@jannisborn I really like the top tokens idea. It was something that I didn't add but it is the natural way to speed up. One thing, I don't see the change. Maybe you forgot to push the latest commits?

**Comment by jannisborn** [comment]
@drugilsberg 
I now refactored the code a lot so that `generate_from_latent` understands the different decoder searches. The generator can now understand all 3 flavors of decoders simply by setting `decoder_search` in the parameter json file to one of `{"greedy", "sampling", "beam"}`.

We can attempt to merge, however the line I really do not like is this:
https://github.com/PaccMann/paccmann_chemistry/blob/a91c938f480ff75720305f03ebb3d6473df60679/paccmann_chemistry/models/vae.py#L260

I am not 100% sure but I think this makes the beam search effectively collapsing to a greedy search. We only take the top1 token to build up the hidden state of the GRU. I had several ideas to circumvent this.

**Option 1:** One would be to keep track of all hidden states and stacks. Very roughly sth like:
```
for t in range(generate_len):
    if t==0:
        output, hidden, stack = self(input_token, hidden, stack)
        top_idx, beams = self.search.step(output, beams)
    elif t==1:
        outputs, hiddens, stacks = zip(*[self(token_id, hidden, stack) for token_id in top_idx])
        top_idxs, beamss = zip(*[self.search.step(output, beams) for output in outputs])
        top_idx, beams, state_ids = # prune the beams to keep only beam_width candidates
    else:
        outputs, hiddens, stacks = zip(*[self(token_id, hiddens[state_id], stacks[state_id]) for token_id,state_id in zip(top_idx, state_ids)])
        top_idxs, beamss = zip(*[self.search.step(output, beams) for output in outputs])
        top_idx, beams, state_ids = # prune the beams to keep only beam_width candidates
 ```

**Option 2:** We try to abuse the batch dimension to track the different hidden states. I am not sure whether this can work in practice, would need to check whether we change the batch size in different calls of the forward function of `stack_rnn`.  Let's assume it works, then at `t=0` we just pass `bs` samples, at `t=1` we pass `beam_width * bs` samples where each of the `beam_width` (aka `top_k` tokens) is used with each corresponding sample. Coding wise this may look more elegant, but we need to prune the discarded hidden states away and only feed the selected ones, so this may be a bit tricky.

What do you think @drugilsberg @niladell ?
Maybe I'm missing sth really simple, would be happy to hear any thoughts.

**Comment by drugilsberg** [comment]
Hi @jannisborn, I think the only issue is that the current code is not performing a pass for each of the beam in the decoder as we discussed. This make it collapse to a greedy search. I don't think we need any trick. We just have to adapt the `step` call to take logits with an additional dimension that is the beam size, and use them in the score update. Now I don't have time, but tomorrow I can work on this.

**Comment by drugilsberg** [comment]
@jannisborn implemented the search with multiple passes. Changed a bunch of signatures to comply with the new implementation. I also removed the search type from the constructor since we might use the same model with different decoding strategies. I also removed the `detach` since the method will be used only in inference. I did not have some good parameters for the model at hand, but we should add a test before merging.

**Comment by jannisborn** [comment]
Great. My conceptual error was a bit stupid namely that if I would do:
```
out1, hidden1, stack1 = stackgru(input_token, hidden, stack)
out2, hidden2, stack2 = stackgru(input_token, hidden, stack)
```
then `out1!=out2` etc. But as this is not the case, I think your solution works. THanks a lot!
Just one comment: We **do need `detach()`**  somewhere, even in `eval` mode (see [here](https://discuss.pytorch.org/t/why-eval-mode-still-requires-detach/37847)).

**Comment by drugilsberg** [comment]
@jannisborn sorry for the `detach` I explained it badly. I would do it in the generator and have the search assuming always detached tensors are passed. So that we control it in the same place for all the strategies.

**Comment by jannisborn** [comment]
@drugilsberg 
Refactored the code a lot, now the decoder searches run smoothly. I think we can attempt a merge into `dev` soon and then implement new beam-search flavors in the future.

**Review by jannisborn — APPROVED** [review]

**Comment by drugilsberg** [comment]
@jannisborn 

I only implemented three fixes:
- keys of the factories are now lower-case to follow the style we have in all other repos
- removed from the package `disable_rdkit_logging` and I moved it only to the example where it's used. In this way `paccmann_chemistry` is pip-installable since it's only relying on PYPI installable packages.
- defaulting to `sampling` search in the example since it's the standard option.

We are good to merge in my opinion.

**Comment by drugilsberg** [comment]
Included changes from dev and merging the PR.
